lib/active_record_shards/migration.rb

```diff
@@ -1,25 +1,32 @@
 # frozen_string_literal: true
 module ActiveRecord
   class Migrator
-    class << self
-      [:up, :down, :run].each do |m|
-        define_method("#{m}_with_sharding") do |*args|
-          ActiveRecord::Base.on_shard(nil) do
-            send("#{m}_without_sharding", *args)
-          end
-          ActiveRecord::Base.on_all_shards do
-            send("#{m}_without_sharding", *args)
-          end
-        end
-        alias_method :"#{m}_without_sharding", m.to_sym
-        alias_method m.to_sym, :"#{m}_with_sharding"
+    def self.migration_context
+      if ActiveRecord::VERSION::STRING >= '5.2.0'
+        ActiveRecord::MigrationContext.new(['db/migrate'])
+      else
+        self
       end
     end
 
+    def run_with_sharding
+      ActiveRecord::Base.on_shard(nil) { run_without_sharding }
+      ActiveRecord::Base.on_all_shards { run_without_sharding }
+    end
+    alias_method :run_without_sharding, :run
+    alias_method :run, :run_with_sharding
+
+    def migrate_with_sharding
+      ActiveRecord::Base.on_shard(nil) { migrate_without_sharding }
+      ActiveRecord::Base.on_all_shards { migrate_without_sharding }
+    end
+    alias_method :migrate_without_sharding, :migrate
+    alias_method :migrate, :migrate_with_sharding
```

Comment:
prefer using a prepended module

Reply:
prepend used to conflict with New Relic’s gem. Dunno if that is still the case.

Anyhow, I think changing our aliasing strategy is out of scope of this PR.

Reply:
Yes let's not change this to prepend as part of this PR.

Reply:
This gem is using the old ruby 1.x style for everything, so I would fix it at scale somewhere else...